File: native-macos/Tests/SeleneNativeTests/LunaFeatureClientTests.swift
import Foundation
import XCTest
@testable import SeleneNative

final class LunaFeatureClientTests: XCTestCase {
    override func tearDown() {
        LunaFeatureTestURLProtocol.requestHandler = nil
        super.tearDown()
    }

    func testShortDramaSearchForwardsCookieAndQuery() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/shortdrama/search")
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(components?.queryValue("query"), "hero")
            XCTAssertEqual(components?.queryValue("page"), "2")
            XCTAssertEqual(components?.queryValue("size"), "24")
            XCTAssertEqual(request.value(forHTTPHeaderField: "Cookie"), "sid=abc")
            return Self.jsonResponse(for: request, body: #"{"data":{"list":[],"total":0}}"#)
        }

        let client = ShortDramaAPIClient(serverURL: URL(string: "http://server.test")!, cookie: "sid=abc", session: makeSession())
        _ = try await client.search(query: "hero", page: 2, pageSize: 24)
    }

    func testShortDramaCategoriesDecodeRawTypeIdArray() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/shortdrama/categories")
            return Self.jsonResponse(for: request, body: #"[{"type_id":7,"type_name":"Hot"}]"#)
        }

        let client = ShortDramaAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let categories = try await client.loadCategories()

        XCTAssertEqual(categories.first?.id, "7")
        XCTAssertEqual(categories.first?.name, "Hot")
    }

    func testShortDramaRecommendDecodesRawItemsArray() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/shortdrama/recommend")
            return Self.jsonResponse(for: request, body: #"[{"id":"s1","name":"Raw Recommend","cover":"c.jpg"}]"#)
        }

        let client = ShortDramaAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let result = try await client.loadRecommend(category: nil, size: 24)

        XCTAssertEqual(result.items.first?.id, "s1")
        XCTAssertEqual(result.total, 1)
    }

    func testShortDramaListUsesCategoryIdQuery() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/shortdrama/list")
            XCTAssertEqual(components?.queryValue("categoryId"), "12")
            XCTAssertNil(components?.queryValue("category"))
            return Self.jsonResponse(for: request, body: #"{"list":[]}"#)
        }

        let client = ShortDramaAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        _ = try await client.loadList(categoryId: "12", page: 1, pageSize: 24)
    }

    func testYouTubePopularUsesRegionAndPageToken() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/youtube/popular")
            XCTAssertEqual(components?.queryValue("regionCode"), "JP")
            XCTAssertEqual(components?.queryValue("pageToken"), "p1")
            return Self.jsonResponse(for: request, body: #"{"items":[],"nextPageToken":"n2"}"#)
        }

        let client = VideoPlatformAPIClient(serverURL: URL(string: "http://server.test")!, cookie: "sid=abc", session: makeSession())
        _ = try await client.loadYouTubePopular(regionCode: "JP", pageToken: "p1")
    }

    func testBilibiliPopularUsesPnPsAndDecodesVideosEnvelope() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/bilibili/popular")
            XCTAssertEqual(components?.queryValue("pn"), "3")
            XCTAssertEqual(components?.queryValue("ps"), "20")
            XCTAssertNil(components?.queryValue("page"))
            return Self.jsonResponse(for: request, body: #"{"videos":[{"id":"BV1","title":"Popular"}]}"#)
        }

        let client = VideoPlatformAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let page = try await client.loadBilibiliPopular(page: 3, pageSize: 20)

        XCTAssertEqual(page.items.first?.id, "BV1")
    }

    func testBilibiliSearchUsesQQuery() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/bilibili/search")
            XCTAssertEqual(components?.queryValue("q"), "music")
            XCTAssertNil(components?.queryValue("query"))
            return Self.jsonResponse(for: request, body: #"{"videos":[]}"#)
        }

        let client = VideoPlatformAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        _ = try await client.searchBilibili(query: "music")
    }

    func testYouTubeSearchUsesQAndDecodesVideosEnvelope() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/youtube/search")
            XCTAssertEqual(components?.queryValue("q"), "trailers")
            XCTAssertNil(components?.queryValue("query"))
            return Self.jsonResponse(for: request, body: #"{"videos":[{"id":"yt1","title":"Trailer"}]}"#)
        }

        let client = VideoPlatformAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let page = try await client.searchYouTube(query: "trailers", contentType: "all", order: "relevance", maxResults: 25)

        XCTAssertEqual(page.items.first?.id, "yt1")
    }

    func testYouTubeRegionsDecodeRegionsEnvelope() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/youtube/regions")
            return Self.jsonResponse(for: request, body: #"{"success":true,"regions":[{"code":"JP","name":"Japan"}]}"#)
        }

        let client = VideoPlatformAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let regions = try await client.loadYouTubeRegions()

        XCTAssertEqual(regions.first?.code, "JP")
    }

    func testDoubanCommentsUseIdPagingSortAndCookie() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/douban/comments")
            XCTAssertEqual(components?.queryValue("id"), "1292052")
            XCTAssertEqual(components?.queryValue("start"), "0")
            XCTAssertEqual(components?.queryValue("limit"), "10")
            XCTAssertEqual(components?.queryValue("sort"), "new_score")
            XCTAssertEqual(request.value(forHTTPHeaderField: "Cookie"), "sid=abc")
            return Self.jsonResponse(for: request, body: #"{"code":200,"data":{"comments":[],"start":0,"limit":10,"total":0}}"#)
        }

        let client = MetadataEnhancementAPIClient(serverURL: URL(string: "http://server.test")!, cookie: "sid=abc", session: makeSession())
        _ = try await client.loadDoubanComments(id: "1292052", start: 0, limit: 10, sort: "new_score")
    }

    func testTmdbBackdropUsesOriginalTitleAndStypeAndDecodesDataEnvelope() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            let components = URLComponents(url: request.url!, resolvingAgainstBaseURL: false)
            XCTAssertEqual(request.url?.path, "/api/tmdb/backdrop")
            XCTAssertEqual(components?.queryValue("original_title"), "Original")
            XCTAssertEqual(components?.queryValue("stype"), "movie")
            XCTAssertNil(components?.queryValue("originalTitle"))
            XCTAssertNil(components?.queryValue("sourceType"))
            return Self.jsonResponse(for: request, body: #"{"data":{"backdrop":"b.jpg","logo":"l.png","poster":"p.jpg"}}"#)
        }

        let client = MetadataEnhancementAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let backdrop = try await client.loadBackdrop(title: "Title", originalTitle: "Original", year: "2026", sourceType: "movie")

        XCTAssertEqual(backdrop?.backdropUrl, "b.jpg")
        XCTAssertEqual(backdrop?.logoUrl, "l.png")
        XCTAssertEqual(backdrop?.posterUrl, "p.jpg")
    }

    func testDoubanRecommendsDecodeListEnvelope() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/douban/recommends")
            return Self.jsonResponse(for: request, body: #"{"code":200,"message":"ok","list":[{"id":"r1","title":"Related","poster":"","year":"2026"}]}"#)
        }

        let client = MetadataEnhancementAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let recommends = try await client.loadDoubanRecommends(kind: "movie", limit: 20, start: 0)

        XCTAssertEqual(recommends.first?.id, "r1")
    }

    func testDoubanQuickInfoMapsRateAndPlotSummary() async throws {
        LunaFeatureTestURLProtocol.requestHandler = { request in
            XCTAssertEqual(request.url?.path, "/api/douban/quick-info")
            return Self.jsonResponse(for: request, body: #"{"data":{"id":"1292052","title":"Title","rate":"9.7","plot_summary":"Summary"}}"#)
        }

        let client = MetadataEnhancementAPIClient(serverURL: URL(string: "http://server.test")!, session: makeSession())
        let info = try await client.loadDoubanQuickInfo(id: "1292052")

        XCTAssertEqual(info?.rating, "9.7")
        XCTAssertEqual(info?.summary, "Summary")
    }

    private func makeSession() -> URLSession {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.protocolClasses = [LunaFeatureTestURLProtocol.self]
        return URLSession(configuration: configuration)
    }

    private static func jsonResponse(for request: URLRequest, body: String) -> (HTTPURLResponse, Data) {
        let response = HTTPURLResponse(url: request.url!, statusCode: 200, httpVersion: nil, headerFields: ["Content-Type": "application/json"])!
        return (response, Data(body.utf8))
    }
}

private final class LunaFeatureTestURLProtocol: URLProtocol {
    static var requestHandler: ((URLRequest) throws -> (HTTPURLResponse, Data))?

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        guard let handler = Self.requestHandler else {
            client?.urlProtocol(self, didFailWithError: APIError.invalidURL)
            return
        }
        do {
            let (response, data) = try handler(request)
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
        } catch {
            client?.urlProtocol(self, didFailWithError: error)
        }
    }

    override func stopLoading() {}
}

private extension URLComponents {
    func queryValue(_ name: String) -> String? {
        queryItems?.first { $0.name == name }?.value
    }
}

File: native-windows/tests/SeleneNative.Tests/ExtendedContent/LunaFeatureClientTests.cs
using System.Net;
using System.Text;
using System.Text.Json;
using SeleneNative.Core.Models;
using SeleneNative.Core.Services;
using Xunit;

namespace SeleneNative.Tests.ExtendedContent;

public sealed class LunaFeatureClientTests
{
    [Fact]
    public async Task ShortDramaClient_SearchAsync_RequestsExpectedPathAndCookie()
    {
        using var handler = new RecordingHandler("""{"data":{"list":[],"total":0}}""");
        var client = new ShortDramaClient("http://server.test", "sid=abc", new HttpClient(handler));

        await client.SearchAsync("hero", page: 2, pageSize: 24);

        Assert.Equal("http://server.test/api/shortdrama/search?query=hero&page=2&size=24", handler.Request!.RequestUri!.ToString());
        Assert.Equal("sid=abc", handler.Request.Headers.GetValues("Cookie").Single());
    }

    [Fact]
    public async Task VideoPlatformClient_LoadYouTubePopularAsync_RequestsRegionAndToken()
    {
        using var handler = new RecordingHandler("""{"items":[],"nextPageToken":"n2"}""");
        var client = new VideoPlatformClient("http://server.test", "sid=abc", new HttpClient(handler));

        await client.LoadYouTubePopularAsync("JP", "p1");

        Assert.Equal("http://server.test/api/youtube/popular?regionCode=JP&pageToken=p1", handler.Request!.RequestUri!.ToString());
    }

    [Fact]
    public async Task VideoPlatformClient_LoadBilibiliPopularAsync_UsesPnPsAndDecodesVideos()
    {
        using var handler = new RecordingHandler("""{"videos":[{"bvid":"BV1","title":"Popular","pic":"p.jpg","play":12345,"pubdate":1710000000}]}""");
        var client = new VideoPlatformClient("http://server.test", httpClient: new HttpClient(handler));

        var page = await client.LoadBilibiliPopularAsync(page: 3, pageSize: 20);

        Assert.Equal("http://server.test/api/bilibili/popular?pn=3&ps=20", handler.Request!.RequestUri!.ToString());
        Assert.Equal("BV1", page.Items.Single().Id);
        Assert.Equal("p.jpg", page.Items.Single().Thumbnail);
        Assert.Equal("12345", page.Items.Single().Views);
        Assert.Equal("1710000000", page.Items.Single().PublishedAt);
    }

    [Fact]
    public async Task VideoPlatformClient_SearchBilibiliAsync_UsesQParameter()
    {
        using var handler = new RecordingHandler("""{"videos":[]}""");
        var client = new VideoPlatformClient("http://server.test", httpClient: new HttpClient(handler));

        await client.SearchBilibiliAsync("music");

        Assert.Equal("http://server.test/api/bilibili/search?q=music", handler.Request!.RequestUri!.ToString());
    }

    [Fact]
    public async Task VideoPlatformClient_SearchYouTubeAsync_UsesQAndDecodesRawApiItems()
    {
        using var handler = new RecordingHandler(
            """
            {"videos":[{"id":{"videoId":"yt1","kind":"youtube#video"},"snippet":{"title":"Trailer","description":"Desc","channelTitle":"Studio","publishedAt":"2026-01-02T00:00:00Z","thumbnails":{"default":{"url":"small.jpg"},"high":{"url":"large.jpg"}}}}],"nextPageToken":"n2"}
            """);
        var client = new VideoPlatformClient("http://server.test", httpClient: new HttpClient(handler));

        var page = await client.SearchYouTubeAsync("trailers", contentType: "video", order: "date", maxResults: 10);

        Assert.Equal("http://server.test/api/youtube/search?q=trailers&contentType=video&order=date&maxResults=10", handler.Request!.RequestUri!.ToString());
        var item = Assert.Single(page.Items);
        Assert.Equal("yt1", item.Id);
        Assert.Equal("Trailer", item.Title);
        Assert.Equal("large.jpg", item.Thumbnail);
        Assert.Equal("Studio", item.Author);
        Assert.Equal("2026-01-02T00:00:00Z", item.PublishedAt);
        Assert.Equal("n2", page.NextPageToken);
    }

    [Fact]
    public void VideoPlatformItem_DeserializesBackwardCompatibleTopLevelShape()
    {
        var json = """{"items":[{"id":"legacy","title":"Legacy","cover":"cover.jpg","thumbnail":"thumb.jpg","views":"99"}]}""";

        var page = JsonSerializer.Deserialize<VideoPlatformPage>(json, new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

        var item = Assert.Single(page!.Items);
        Assert.Equal("legacy", item.Id);
        Assert.Equal("Legacy", item.Title);
        Assert.Equal("cover.jpg", item.Cover);
        Assert.Equal("thumb.jpg", item.Thumbnail);
        Assert.Equal("99", item.Views);
    }

    [Fact]
    public async Task ShortDramaClient_LoadShortDramaListAsync_UsesCategoryIdParameter()
    {
        using var handler = new RecordingHandler("""{"list":[]}""");
        var client = new ShortDramaClient("http://server.test", httpClient: new HttpClient(handler));

        await client.LoadShortDramaListAsync("12", page: 2, pageSize: 24);

        Assert.Equal("http://server.test/api/shortdrama/list?categoryId=12&page=2&size=24", handler.Request!.RequestUri!.ToString());
    }

    [Fact]
    public async Task MetadataEnhancementClient_LoadBackdropAsync_UsesLunaTvParameterNames()
    {
        using var handler = new RecordingHandler("""{"data":{"backdropUrl":"b.jpg"}}""");
        var client = new MetadataEnhancementClient("http://server.test", httpClient: new HttpClient(handler));

        await client.LoadBackdropAsync("Title", originalTitle: "Original", year: "2026", sourceType: "movie");

        Assert.Equal("http://server.test/api/tmdb/backdrop?title=Title&original_title=Original&year=2026&stype=movie", handler.Request!.RequestUri!.ToString());
    }

    [Fact]
    public async Task MetadataEnhancementClient_LoadDoubanCommentsAsync_RequestsIdStartLimitSort()
    {
        using var handler = new RecordingHandler("""{"code":200,"data":{"comments":[],"start":0,"limit":10,"total":0}}""");
        var client = new MetadataEnhancementClient("http://server.test", "sid=abc", new HttpClient(handler));

        await client.LoadDoubanCommentsAsync("1292052", start: 0, limit: 10, sort: "new_score");

        Assert.Equal("http://server.test/api/douban/comments?id=1292052&start=0&limit=10&sort=new_score", handler.Request!.RequestUri!.ToString());
    }

    private sealed class RecordingHandler(string json) : HttpMessageHandler, IDisposable
    {
        public HttpRequestMessage? Request { get; private set; }

        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request,
            CancellationToken cancellationToken)
        {
            Request = request;
            return Task.FromResult(new HttpResponseMessage(HttpStatusCode.OK)
            {
                Content = new StringContent(json, Encoding.UTF8, "application/json")
            });
        }
    }
}
